### ISSUE TITLE
apiserver: panic from goroutine spun up by request handler should not crash the apiserver

### ISSUE THREAD

**Issue by tkashem** [kubernetes/kubernetes#115565]
### What happened?

request handler logic spins up goroutine, and most of them has the style: 
```
go dunc() {
     defer utilruntime.HandleCrash()

    ...
}
```

`HandleCrash` will re-raise the panic after handling it, if `ReallyCrash` is true. (`ReallyCrash` is true by default, and to my knowledge it's not overridden to `false`)

https://github.com/kubernetes/kubernetes/blob/2ab3151936f1df6b3ffb4d49bf9dc90218264545/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L54-L57
The re-raised panic will go unhandled and cause the process to crash.

I have found the following code sites:
https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L289

https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1173-L1175

https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1192-L1194

the following may not be request scoped, but if it panics do we want the apiserver to crash?
https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1444

**Comment by liggitt** [comment]
> request handler logic spins up goroutine

I agree that anything creating goroutines in the request handling chain should catch/propagate panics upward so they are caught by the panic-handling filter and terminate the request but not the server. For what it's worth, not everything you linked is in the request handling chain, so some of those are probably appropriate to exit on if they panic, like this one:

https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L478

I also agree that HandleCrash re-raising the panic and continuing to crash is confusing. When that behavior was added in https://github.com/kubernetes/kubernetes/pull/28800, it essentially made HandleCrash unsuitable for use within the request handling chain.

**Comment by tkashem** [comment]
I removed it from the issue description
https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L478

> I agree that anything creating goroutines in the request handling chain should catch/propagate panics upward so they are caught by the panic-handling filter

I will work on https://github.com/kubernetes/kubernetes/pull/115564 and prepare it for review, thank you for your feedback!

**Comment by aojea** [comment]
golang http server can recover from panic by default https://go.dev/play/p/NWeR5o-AkZq

https://github.com/golang/go/issues/16542, should this not solve the problem?

**Comment by tkashem** [comment]
i had a work in progress util package that provides the abstraction to recover and propagate panics from a child goroutine to the parent.
https://github.com/kubernetes/kubernetes/pull/115564

**Comment by MikeSpreitzer** [comment]
In thinking about this, I noticed that `func (e *Store) DeleteCollection` has some concurrency bugs.

 The distributor goroutine could stuff the last `2*workersNumber` indices in `toProcess` before any worker attempts to work on those, and later all the workers could suffer a panic. The intended warning from https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1181 will not happen in this case. Also, https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1193 is not guaranteed to happen before https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1217 , so the `select` at https://github.com/kubernetes/kubernetes/blob/1f02f43ec6910cbcd46222b16a2f3965458e5e45/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L1219-L1224 is not guaranteed to take the `err := <-errs` case.

**Comment by MikeSpreitzer** [comment]
@MikeSpreitzer
